### PR TITLE
receiver/dockerstatsreceiver: add container.uptime metric

### PR DESCRIPTION
**Pull request by gbbr**
This change adds the `container.uptime` metric which indicates time elapsed (in seconds) since the container started.

Updates #22037

### CODE REVIEW

File: receiver/dockerstatsreceiver/receiver.go [hunk 1]
```diff
@@ -264,3 +266,14 @@ func (r *receiver) recordPidsMetrics(now pcommon.Timestamp, pidsStats *dtypes.Pi
 		}
 	}
 }
+
+func (r *receiver) recordBaseMetrics(now pcommon.Timestamp, base *types.ContainerJSONBase) {
+	t, err := time.Parse(time.RFC3339, base.State.StartedAt)
+	if err != nil {
+		// value not available or invalid
```
Comment (mx-psi) [inline receiver/dockerstatsreceiver/receiver.go:273]: This sounds like something that should be bubbled up as a [partial scraper error](https://pkg.go.dev/go.opentelemetry.io/collector/receiver/scrapererror#PartialScrapeError)

File: receiver/dockerstatsreceiver/receiver_test.go [hunk 2]
```diff
@@ -253,11 +256,40 @@ func TestScrapeV2(t *testing.T) {
 
 			assert.NoError(t, err)
 			assert.NoError(t, pmetrictest.CompareMetrics(expectedMetrics, actualMetrics,
-				pmetrictest.IgnoreResourceMetricsOrder(), pmetrictest.IgnoreStartTimestamp(), pmetrictest.IgnoreTimestamp()))
+				pmetrictest.IgnoreResourceMetricsOrder(),
+				pmetrictest.IgnoreStartTimestamp(),
+				pmetrictest.IgnoreTimestamp(),
+				pmetrictest.IgnoreMetricValues(
+					"container.uptime", // value depends on time.Now(), making it unpredictable as far as tests go
```
[the remark this answers is not in the document]
Comment (gbbr) [reply receiver/dockerstatsreceiver/receiver_test.go:263]: I think that would be [pretty tricky](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/be6d97154dbf826eeef39ddb48915d8d6f7e6188/receiver/dockerstatsreceiver/receiver.go#L101) and I'm not sure if it's worth the effort. I added a separate test to make sure that the values are correct. Do you think that's not enough?

File: receiver/dockerstatsreceiver/receiver.go (hunk 1 above)
Comment (gbbr) [reply receiver/dockerstatsreceiver/receiver.go:279]: Done.

File: receiver/dockerstatsreceiver/metadata.yaml [hunk 3]
```diff
@@ -639,3 +639,11 @@ metrics:
       value_type: int
       aggregation: cumulative
       monotonic: false
+
+  # Base
+  container.uptime:
+    enabled: true
+    description: "Time elapsed since container start time."
+    unit: s
```
[the remark this answers is not in the document]
Comment (gbbr) [reply receiver/dockerstatsreceiver/metadata.yaml:647]: There are [other metrics](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/be6d97154dbf826eeef39ddb48915d8d6f7e6188/receiver/hostmetricsreceiver/internal/scraper/diskscraper/metadata.yaml#L47) that have second units, despite not being in this receiver. Isn't nanoseconds too precise for something like uptime?

[the remark this answers is not in the document]
Comment (gbbr) [reply receiver/dockerstatsreceiver/metadata.yaml:647]: Most uptime metrics seem to be seconds ([1](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3d1123220122ab0498324f1fcac77453e0ea4200/receiver/iisreceiver/testdata/scraper/expected.yaml#L152-L153), [2](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3d1123220122ab0498324f1fcac77453e0ea4200/receiver/apachereceiver/metadata.yaml#L60-L63), [3](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3d1123220122ab0498324f1fcac77453e0ea4200/receiver/redisreceiver/testdata/integration/expected.yaml#L243-L251), [4](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3d1123220122ab0498324f1fcac77453e0ea4200/receiver/saphanareceiver/testdata/expected_metrics/full.yaml#L769-L783)), so I think this is fine.

However, just my 2 cents: I don't think metrics need to be consistent per-receiver cross-concern. CPU time is one thing, and uptime is another (where ns precision might be overkill).

**Review by mx-psi — APPROVED** [review]

**Review by jamesmoessis — APPROVED** [review]
LGTM, thanks!

File: receiver/dockerstatsreceiver/metadata.yaml [hunk 4]
```diff
@@ -639,3 +639,11 @@ metrics:
       value_type: int
       aggregation: cumulative
       monotonic: false
+
+  # Base
+  container.uptime:
+    enabled: true
```
Comment (atoulme) [inline receiver/dockerstatsreceiver/metadata.yaml:645]: Should this be disabled by default?

Comment (gbbr) [reply receiver/dockerstatsreceiver/metadata.yaml:645]: I don't know. You tell me 🙂 I figured that as a "base" metric I'd have it enabled, but I'm ok with either.

Comment (atoulme) [reply receiver/dockerstatsreceiver/metadata.yaml:645]: When a new metric is added, it's better if it's marked as not enabled by default so we avoid ingesting new metrics on upgrade by mistake. We can turn it on by default in a separate PR.

Comment (atoulme) [reply receiver/dockerstatsreceiver/metadata.yaml:645]: See here for more context: https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/scraping-receivers.md#adding-a-new-default-metric-or-turning-an-existing-optional-metric-into-default

Comment (gbbr) [reply receiver/dockerstatsreceiver/metadata.yaml]: Great. Thanks a lot for jumping in and clarifying. I disabled it. PTAL

Comment (gbbr) [reply receiver/dockerstatsreceiver/metadata.yaml]: @atoulme would you mind taking another look please?

**Review by songy23 — APPROVED** [review]

**Comment by mx-psi** [comment]
@atoulme is this good to go from your side?

**Comment by mx-psi** [comment]
needs a rebase

**Review by atoulme — APPROVED** [review]
LGTM!

**Pull request merged**